### PR TITLE
perf(markdown): reduce streaming stalls with split rendering

### PR DESCRIPTION
**Pull request by zhangmo8**
## Summary

Reduces the renderer stalls and dropped frames on long markdown streams, which was the underlying cost behind the frame-rate complaint in #2174.

The streaming markdown renderer previously re-rendered the **whole document** on every token/commit, so long streams blocked the renderer for ~1 s at a time (worst measured stall ~1.9 s) with heavy GC churn. Two changes bound that cost:

### 1. Length-adaptive update coalescing
Stream updates are debounced more aggressively as the document grows (was a fixed 32 ms), so the whole-document re-parse happens less often on long streams.

### 2. Committed-prefix + live-tail split render
Once the stream exceeds ~6 KB, the content renders as two segments:

- a **static committed prefix** (re-rendered only when the split advances every ~6 KB, with markdown node virtualization limiting the cost), and
- a small **live streaming tail** (≤ ~6 KB) that keeps the smooth typewriter-style reveal.

Each token now re-renders only the small tail instead of the whole document:

```
BEFORE                              AFTER
one renderer for the whole doc      [static committed prefix (virtualized)]
every token -> full re-render       + [small live tail, smooth reveal]
O(document) per token               O(tail) per token
```

### Measured (DevTools performance profiles, macOS)
- Smooth-stream `tick`: ~1 s → **max ~6 ms**
- Worst renderer stall: ~1.9 s → **~100–300 ms** (split-advance only)
- Idle and short streams: unchanged behavior
- The typewriter reveal stays continuous (~25–57 content updates/s)

## Notes
- No user-facing FPS-limit setting is added: `webContents.setFrameRate()` only applies to offscreen rendering (electron/electron#22873).
- Refs: #2174

## Testing
- `typecheck`, `oxlint`, `oxfmt` clean.
- `MarkdownRenderer` renderer tests pass (21/21).
- Verified with performance profiles: renderer long tasks and dropped frames drop materially on long streaming responses.

## Summary by CodeRabbit

* **New Features**
  * Improved rendering performance for long, streaming Markdown documents.
  * Separated stable content from the actively streaming portion for smoother updates.
  * Preserved code blocks and paragraph formatting by splitting only at safe boundaries.
  * Applied independent rendering and virtualization limits to each document segment.
  * Adjusted update timing dynamically based on document length for a more responsive experience.
  * Reset streaming state cleanly when document generation finishes.

### CODE REVIEW

**Review by zerob13 — CHANGES_REQUESTED** [review]
## Code Review — split rendering for streaming stalls

The split-render approach (committed prefix + live tail) is a sound direction for reducing streaming stalls on long documents, and the adaptive `parseCoalesceMs` levels are reasonable. A few issues should be addressed before merging.

---

### 1. `committedLength` not reset when content shrinks (bug)

**File:** `MarkdownRenderer.vue:320-325`

When a stream is interrupted and regenerated, `props.content` may reset to a shorter value while `isStreaming` stays `true`. `committedLength` is only reset on the `isStreaming` `true → false` transition (line 327-329), not on content shrink. This leaves `committedLength` at a stale large value:

- `committedContent` = `renderContent.slice(0, staleLength)` → returns entire content
- `tailContent` = `renderContent.slice(staleLength)` → returns empty string
- Split render silently breaks until content grows past `staleLength + STREAM_TAIL_CAP_CHARS`

**Fix:** add a shrink guard:

```ts
watch(renderContent, (content) => {
  if (!isStreaming.value) return
  if (content.length < committedLength.value) {
    committedLength.value = 0
  }
  if (content.length - committedLength.value > STREAM_TAIL_CAP_CHARS) {
    committedLength.value = findSafeSplit(content, content.length - STREAM_MIN_TAIL_CHARS)
  }
})
```

### 2. `findSafeSplit` can leave committed prefix inside an unclosed fence (correctness)

**File:** `MarkdownRenderer.vue:290-308`

When the prefix has an unmatched fence and `beforeFence < minSplit`, the fallback `split` stays at `maxSplit` — inside the unclosed fence. The prefix segment renders an unclosed code block while the tail re-opens the same fence, causing the code block to appear duplicated.

**Fix:** when no safe boundary is found, don't advance past the current committed position:

```ts
if (fenceCount % 2 === 1) {
  // ...existing logic...
  if (beforeFence >= minSplit) {
    split = Math.min(maxSplit, beforeFence + 2)
  } else {
    return committedLength.value  // don't advance into unclosed fence
  }
}
```

### 3. `RenderSegment.streaming` is dead code

**File:** `MarkdownRenderer.vue:335, 358, 399`

The `streaming` field is declared in `RenderSegment`, assigned in both segments, but the template (lines 6-34) never binds it to `NodeRenderer`. Remove the field, or bind it if `NodeRenderer` accepts a `streaming` prop.

### 4. Comment references non-existent function

**File:** `MarkdownRenderer.vue:234`

```
// (see splitRenderContent below)
```

There is no `splitRenderContent`. Should reference `findSafeSplit` or `renderSegments`.

### 5. `streamUpdateDebouncers` naming implies streaming usage but is dead during streaming (clarity)

**File:** `MarkdownRenderer.vue:531-557`

`commitImmediately` is `true` whenever `isStreaming` is `true` (line 561), so all streaming updates bypass the debouncers. This matches the original design ("Main already coalesces renderer snapshots"), but `STREAM_UPDATE_COALESCE_LEVELS` / `streamUpdateDebouncers` naming implies they handle streaming. Only `parseCoalesceMs` from the levels is actually consumed during streaming (via the segment prop).

Consider renaming (e.g. `CONTENT_UPDATE_COALESCE_LEVELS`) or adding a comment clarifying the debounce path is for non-streaming content updates only.

---

### Minor observations

- **DOM structure change:** single `NodeRenderer` → `v-for` multiple segments. Worth verifying no consumers depend on a single child under `.markdown-renderer-root`.
- **No tests:** split render is a significant rendering logic change; a basic regression test verifying split/merge output would help.
- **`parseCoalesceMs` uses `props.content.length` while `usingSplitRender` uses `renderContent.value.length`:** pre-normalization vs post-normalization — usually a small delta but could cause level mismatch in edge cases.

File: src/renderer/src/components/markdown/MarkdownRenderer.vue [hunk 1]
```diff
@@ -241,9 +265,155 @@ const renderBatchBudgetMs = computed(() =>
...
+  if (fenceCount % 2 === 1) {
+    let opener = -1
+    for (const position of fenceLineStarts(content)) {
+      if (position < split) opener = position
+    }
+    const beforeFence = opener >= 0 ? content.lastIndexOf('\n\n', opener) : -1
+    if (beforeFence >= minSplit) split = Math.min(maxSplit, beforeFence + 2)
+  }
+  return split
+}
```
[the remark this answers is not in the document]
Comment (zhangmo8) [reply src/renderer/src/components/markdown/MarkdownRenderer.vue]: Fixed. `findSafeSplit` now returns the current committed length (no advance) when no fence-safe boundary exists, so a single fenced block larger than the tail cap stays entirely in the streaming tail until it closes — the static prefix never contains an unterminated code block.

File: src/main/desktop/window/index.ts [hunk 2]
```diff
@@ -653,11 +654,9 @@ export class WindowPresenter implements IWindowPresenter {
...
       icon: iconFile, // 设置图标
       titleBarStyle: process.platform === 'darwin' ? 'hiddenInset' : undefined, // macOS 风格标题栏
-      transparent: process.platform === 'darwin', // macOS 标题栏透明
-      vibrancy: process.platform === 'darwin' ? 'under-window' : undefined, // macOS 磨砂效果
-      visualEffectState: process.platform === 'darwin' ? 'followWindow' : undefined,
-      backgroundMaterial: process.platform === 'win32' ? 'mica' : undefined, // Windows 11 材质效果
-      backgroundColor: '#00ffffff', // 透明背景色
+      // 不透明纯色背景：去掉透明/磨砂/Mica 材质，避免 OS 层持续合成导致的
+      // 前台高帧率与 GPU 空转。颜色跟随系统主题，与渲染层背景保持一致。
+      backgroundColor: nativeTheme.shouldUseDarkColors ? '#0d0d0d' : '#ffffff',
```
[the remark this answers is not in the document]
Comment (zhangmo8) [reply src/main/desktop/window/index.ts]: No longer applicable: the window-background change this comment targeted (`src/main/desktop/window/index.ts`) was dropped from this PR. Only the streaming-markdown change remains.

**Comment by zhangmo8** [comment]
Review feedback addressed in 6598e1ac5:

- **RenderSegment.streaming unused** → removed the field
- **Coalescing routing** → `STREAM_UPDATE_COALESCE_LEVELS` now routes on `isStreaming` instead of `props.smoothStreaming`, so the stream policy never applies to static documents
- **Fence-safe split** → `findSafeSplit` no longer advances into an unterminated fenced block; it keeps the current split until the fence closes
- **Window-background comment** → not applicable, that change was dropped from this PR

**Comment by zhangmo8** [comment]
Thanks for the detailed review @zerob13 — all points addressed in 4a867433d:

1. **`committedLength` not reset on shrink** — fixed: the `renderContent` watcher now resets `committedLength` to 0 whenever the content shrinks below it, and runs `immediate` so a long stream arriving at mount also splits correctly on first paint. (Also covered by a new regression test.)
2. **Split inside an unclosed fence** — fixed (also flagged by CodeRabbit): `findSafeSplit` returns the current committed length when no fence-safe boundary exists, and `usingSplitRender` now requires `committedLength > 0`, so an unsplittable stream (e.g. a single fenced block > tail cap) renders as one streaming document instead of a broken prefix.
3. **`RenderSegment.streaming` dead field** — removed.
4. **Stale `splitRenderContent` comment** — now references `findSafeSplit`/`renderSegments`.
5. **Naming clarity** — renamed `STREAM_UPDATE_COALESCE_LEVELS`/`streamUpdateDebouncers` to `CONTENT_UPDATE_COALESCE_LEVELS`/`contentUpdateDebouncers`; the comment now explains streaming updates bypass the debouncers via `commitImmediately` and only `parseCoalesceMs` is consumed live.

Misc:
- **DOM structure**: verified no consumer depends on a single child under `.markdown-renderer-root`; the root keeps its classes and handlers.
- **Tests**: added 3 regression tests (long-stream split into prefix/tail, oversized-fence no-split, shrink-reset).
- **`parseCoalesceMs` source**: aligned to `renderContent.value.length` so the coalescing level and the split decision use the same (post-normalization) length.

**Comment by zhangmo8** [comment]
Both remaining threads fixed in 05866cd1c (and resolved):

- **Do not advance the split when no blank-line boundary exists** — `findSafeSplit` now returns the current committed length when no `

` is found in the window, so a long paragraph/list is never chopped mid-way (rendered as one streaming document until a safe boundary appears).
- **Fence detection mixed-marker parity** — `fenceLineStarts` now tracks `` ` `` and `~~~` fence lines separately and the split guard balances each marker type independently.

Added regression tests for both cases (26 tests total, all passing locally).

**Review by zerob13 — APPROVED** [review]
## Re-review — APPROVED ✅

All 5 findings from the previous CHANGES_REQUESTED review are resolved. No new issues found.

### Previous findings — all fixed

| # | Finding | Status |
|---|---------|--------|
| 1 | `committedLength` not reset when content shrinks | ✅ Fixed — guard added at `watch(renderContent)`: `if (content.length < committedLength.value) committedLength.value = 0` |
| 2 | `findSafeSplit` fallback leaves committed prefix inside unclosed fence | ✅ Fixed — returns `current` (no advance) when no safe blank-line boundary exists or fence is unbalanced |
| 3 | `RenderSegment.streaming` field never bound to template | ✅ Fixed — field removed entirely |
| 4 | Comment references non-existent `splitRenderContent` | ✅ Fixed — now correctly references `findSafeSplit / renderSegments` |
| 5 | `streamUpdateDebouncers` dead during streaming, naming misleading | ✅ Fixed — renamed to `CONTENT_UPDATE_COALESCE_LEVELS`, comment clarifies streaming bypasses debouncers via `commitImmediately` |

### New tests — reasonable coverage

5 new test cases covering all boundary conditions from the previous review:
- Split long streaming content into prefix + tail
- No split inside a single oversized code fence
- No split when mixed fence markers (`\`\`\`` vs `~~~`) are unbalanced
- No split mid-paragraph when no blank-line boundary exists
- Reset split when stream shrinks and restarts

All 26 tests pass (21 existing + 5 new). No over-testing.

### Nit (non-blocking)

`findSafeSplit` line 303: the `else` branch (`: maxSplit`) is dead code — `if (blankLine < minSplit) return current` on the next line makes it unreachable. Could simplify to:
```ts
if (blankLine < minSplit) return current
let split = Math.min(maxSplit, blankLine + 2)
```

Not blocking — feel free to clean up in a follow-up.

### Design assessment

- **Over-engineering**: No. Two-segment split (prefix + tail) is the minimal viable approach. Graceful degradation to single-document render when safe split is impossible.
- **Pattern consistency**: Uses Vue 3 composition API, `useDebounceFn` from VueUse, `computed`/`watch`/`ref` — consistent with existing codebase patterns.
- **Breaking changes**: DOM structure changes from single `NodeRenderer` to `v-for`, but only during streaming and doesn't affect external props/emits interface. Non-streaming path unchanged.
- **Test scope**: 5 tests for boundary conditions — appropriate, not excessive.

Ship it 🚀
